src/oom/oomd-util.c

```diff
@@ -0,0 +1,352 @@
+                                curr_mem_pressure.avg10 * 3 + curr_mem_pressure.avg60 * 2 + curr_mem_pressure.avg300) {
+                        curr_mem_pressure = ctx->memory_pressure;
+                        target = key;
+                }
+
+        if (curr_mem_pressure.avg10 > threshold) {
+                if (*last_hit_memory_threshold == 0)
+                        *last_hit_memory_threshold = now(CLOCK_REALTIME);
+
+                diff = now(CLOCK_REALTIME) - *last_hit_memory_threshold;
+                if (diff >= duration) {
+                        log_debug("Memory pressure for %s is greater than %.1f for more than %"PRIu64" seconds",
+                                        target, threshold, duration / USEC_PER_SEC);
+                        return target;
+                }
+        } else
+                *last_hit_memory_threshold = 0;
+
+        return NULL;
+}
+
+bool oomd_swap_free_below(const OomdSystemContext ctx, unsigned threshold_percent) {
+        uint64_t swap_threshold;
+        bool ret;
+
+        assert(threshold_percent <= 100);
+
+        swap_threshold = ctx.swap_total * ((uint64_t) threshold_percent) / 100;
+        ret = (ctx.swap_total - ctx.swap_used) < swap_threshold;
+        if (ret)
+                log_debug("Swap total (%"PRIu64") - used (%"PRIu64") is less than %u%%",
+                                ctx.swap_total, ctx.swap_used, threshold_percent);
+        return ret;
+}
+
+int compare_pgscan(const OomdCGroupContext *c1, const OomdCGroupContext *c2) {
+        assert(c1);
+        assert(c2);
+
+        if (c1->pgscan > c2->pgscan)
+                return 1;
+        else if (c1->pgscan < c2->pgscan)
+                return -1;
+        else
+                return 0;
+}
+
+int compare_swap_usage(const OomdCGroupContext *c1, const OomdCGroupContext *c2) {
+        assert(c1);
+        assert(c2);
+
+        if (c1->swap_usage > c2->swap_usage)
+                return 1;
+        else if (c1->swap_usage < c2->swap_usage)
+                return -1;
+        else
+                return 0;
+}
+
+int oomd_sort_cgroups(Hashmap *h, compare_func_t compare_func, const char *prefix, char ***ret) {
+        _cleanup_free_ char **sorted_cgroups = NULL;
+        OomdCGroupContext *c1, *c2;
+        Iterator i;
+        char *key;
+
+        assert(h);
+        assert(compare_func);
+        assert(ret);
+
+        HASHMAP_FOREACH_KEY(c1, key, h, i)
+                if (prefix == NULL || startswith(key, prefix)) {
+                        bool inserted = false;
+                        int r, j = 0;
+                        char **cgs;
+
+                        STRV_FOREACH(cgs, sorted_cgroups) {
+                                c2 = hashmap_get(h, *cgs);
+                                assert(c2);
+
+                                if (compare_func(c1, c2) > 0) {
+                                        r = strv_insert(&sorted_cgroups, j, key);
```

Comment:
strv_insert() is really slow, doing one-increments of memory reallocation. This whole function becomes O(n^3) through that, no?

Maybe just put the cgroup paths in an strv (using plain strv_extend() which is more efficient, it grows the size exponentially), and afteraftwards use qsort() on it to order it? That's gonna be a ton faster